![](https://media-exp1.licdn.com/dms/image/C4D0BAQHUQWhqV2rl1g/company-logo_200_200/0/1550502396615?e=2159024400&v=beta&t=_mM0D3cbDu8DL3MUvcb75g65zJ-c4Wd0nrguJGoW_gE)

# Python (programowanie funkcyjne) - funkcje

_Mikołaj Leszczuk_

## Konspekt

* **Programowanie funkcyjne** jest paradygmatem programowania, gdzie pierwsze skrzypce należą do funkcji
* W tym module poznamy podstawy programowania w tej technice w języku Python

## Programowanie funkcyjne

* Filozofia i metodyka programowania będąca odmianą programowania deklaratywnego, w której:
  * Funkcje należą do wartości podstawowych, a 
  * Nacisk kładzie się:
    * Na wartościowanie (często rekurencyjnych) funkcji, a 
    * Nie na wykonywanie poleceń
* W czystym programowaniu funkcyjnym, raz zdefiniowana funkcja zwraca zawsze tę samą wartość dla danych wartości argumentów, tak jak funkcje matematyczne
* Duży nacisk kładzie się także na niezmienność danych

## Historia

* Podstawą teoretyczną programowania funkcyjnego jest opracowany w latach 30. XX wieku przez **Alonzo Churcha rachunek lambda**, a dokładnie rachunek lambda z typami
* Nazwa paradygmatu, po raz pierwszy pojawiła się w pracy **Johna Backusa** „Can Programming Be Liberated From the von Neumann Style? A Functional Style and Its Algebra of Programs” w roku 1977, dzięki której Backus dostał **nagrodę Turinga**, mimo że programowanie funkcyjne było znane już wcześniej

* Pierwszym funkcyjnym językiem programowania był **Information Processing Language** (IPL) opracowany w połowie lat 50. XX wieku dla maszyny JOHNNIAC przez:
  * Allena Newella,
  * Cliffa Shawa, i
  * Herberta Simona
* Jednakże przełom dla programowania funkcyjnego stanowiło opracowanie przez Johna McCarthy pod koniec lat 50. języka **LISP** dla maszyn IBM 700/7000
* **LISP** wprowadził wiele cech spotykanych we współczesnych językach programowania
* Opracowany w latach 70. język **Scheme** miał za zadanie uproszczenie i poprawienie języka LISP

* W latach 70. powstały dwa kolejne funkcyjne języki programowania: 
  * Język **ML** opracowany przez Robina Milnera na Uniwersytecie w Edynburgu, oraz 
  * Język **Miranda** opracowany przez Davida Turnera na Uniwersytecie w Kent
* Język ML dał początek kilku dialektom, z których najpopularniejsze obecnie to:
  * **Objective Caml**, oraz
  * **Standard ML**
* Pod koniec lat 80. został opracowany i ustandaryzowany język programowania **Haskell**, wywodzący się z języka Miranda
* Jednym z najnowszych języków funkcyjnych jest rozwijany przez firmę **Microsoft** język **F#**
* Jest to przeniesienie języka **Objective Caml** na platformę **.NET**

## Podział języków funkcyjnych

* Języki funkcyjne można podzielić na dwie grupy:
  * Języki czysto funkcyjne
  * Języki mieszane

## Języki czysto funkcyjne

* Do tej grupy należą języki, w których nie występują zmienne ani **efekty uboczne**
* Wejście/wyjście w takich językach musi się odbywać jakimś alternatywnym mechanizmem, np. za pomocą **monad**
* Przedstawiciele tej podgrupy to **Haskell** oraz **Clean**

## Języki mieszane

* Języki tej grupy są popularniejsze niż języki czysto funkcyjne, gdyż:
  * Umożliwiają one stosowanie zmiennych
  * Pozwalają na:
    * Efekty uboczne,
    * Tradycyjne wejście/wyjście, i
    * Mieszanie stylu funkcyjnego z:
      * **Imperatywnym**, bądź
      * **Obiektowym**
* Wartościowanie w nich jest przeważnie **zachłanne**

* Do grupy tej należą:
  * **Lisp** z wszystkimi pochodnymi (np. **Clojure**, **Scheme**), **Erlang**, **Scala**
  * Języki grupy **ML** (**Standard ML**, **OCaml**, więc i oparty na nim **F#**)
  * Język **Nemerle** (tworzony przez polskich studentów)
* Ponadto elementy programowania funkcyjnego występują również w językach takich jak **Java** (od wersji 8), **Python**, **Ruby**
* A nawet do pewnego stopnia **Perl**, **JavaScript**, **D**

## Funkcje anonimowe (lambda)

* Python umożliwia tworzenie funkcji w miejscu, bez osobnego deklarowania ich
* Takie funkcje nazywane są „anonimowe”, bo:
  * Nie muszą przyjmować nazw, a
  * Wykorzystuje się je często jako np. argumenty do „normalnych” funkcji
* Podstawowa składania funkcji anonimowej jest następująca:

```
lambda arg1, arg2: arg1 ** arg2
       \--------/  \----------/
        argumenty   operacja, której wynik jest zwracany przez funkcję
```

#### Podobieństwo do Excela

* **Operacja funkcji `lambda` jest niemal identyczna jak zawartość funkcji w komórce w Excelu**
* **Różnicą jest jedynie odwoływanie do:**
  * **Zmiennych (Python), i**
  * **Nazw komórek (Excel)**

#### Przykład funkcji anonimowej

* Przykład funkcji anonimowej obliczającej przeciwprostokątną w trójkącie prostokątnym (twierdzenie Pitagorasa)
* Przypomnienie:
  * Twierdzenie Pitagorasa: `c*c = a*a + b*b`
  * Operator `**` to podnoszenie do potęgi
  * Jeśli wykładnik potęgi ma postać `1/n` (przykładowo: `1/2`), to takie potęgowanie zamienia się w pierwiastkowanie o stopniu `n`

```python
lambda a, b:  ((a * a) + (b * b)) ** 0.5
```

* Funkcje anonimowe są obiektami w Pythonie
* Dzięki temu możemy je przypisać w taki sam sposób, jak przypisanie wartości do zmiennej

```python
pitagoras = lambda a, b: ((a * a) + (b * b)) ** 0.5
```

* Pozwala to w dalszym kodzie np. wywoływać taką funkcję anonimową po jej, de facto, nazwie:

In [29]:
pitagoras = lambda a, b:  ((a * a) + (b * b)) ** 0.5
print(pitagoras(3, 5))

5.830951894845301


#### Funkcje konwencjonalne vs. lambda

* Funkcja `lambda` może wyglądać jak zabawka dla matematyka 
* Formalizm, który zamienia łatwą do zrozumienia kwestię w abstrakcyjny, trudniejszy do uchwycenia formalizm
* Przede wszystkim mogliśmy osiągnąć ten sam efekt, używając po prostu następującej definicji funkcji konwencjonalnej

In [97]:
def c_sum(x, y):
    return x + y

In [98]:
l_sum = lambda x, y: x + y

In [99]:
print(c_sum(3, 4))
print(l_sum(3, 4))

7
7


* Mogę Was zapewnić, że zalety tego podejścia (`lambda`) będą widoczne, gdy nauczycie się (później) używać funkcji `map()`

#### Ćwiczenie na rozgrzewkę

* Utwórz funkcję `lambda`, która przyjmuje jeden parametr (`a`) i zwraca go

#### Rozwiązanie ćwiczenia na rozgrzewkę

In [1]:
pass

#### Kolejne ćwiczenia na rozgrzewkę

* Napisz program w Pythonie, aby utworzyć funkcję `lambda`, która dodaje `15` do podanej liczby przekazanej jako argument
* A także utwórz funkcję `lambda`, która mnoży argument `x` przez argument `y` i drukuje wynik

#### Rozwiązania kolejnych ćwiczeń na rozgrzewkę

In [2]:
pass

In [3]:
pass

#### Ćwiczenie

* Napisz program w Pythonie do sortowania(względem oceny) listy krotek za pomocą `lambda`

```python
subject_marks = [('Język angielski', 88),
                 ('Nauka',           90),
                 ('Matematyka',      97),
                 ('Nauki społeczne', 82)]
```

#### Przykładowe rozwiązanie

In [4]:
subject_marks = [('Język angielski', 88),
                 ('Nauka',           90),
                 ('Matematyka',      97),
                 ('Nauki społeczne', 82)]
pass

#### Ćwiczenie

* Napisz program w Pythonie, aby posortować listę słowników za pomocą `lambda`

```python
models = [{'marka': 'Nokia',   'model': '3310',   'kolor': 'Czarny'},
          {'marka': 'Apple',   'model': '11',     'kolor': 'Złoty'},
          {'marka': 'Samsung', 'model': 'Galaxy', 'kolor': 'Srebrny'}]
```

#### Przykładowe rozwiązanie

In [105]:
models = [{'marka': 'Nokia',   'model': '3310',   'kolor': 'Czarny'},
          {'marka': 'Apple',   'model': '11',     'kolor': 'Złoty'},
          {'marka': 'Samsung', 'model': 'Galaxy', 'kolor': 'Srebrny'}]
pass

Oryginalna lista słowników :
[{'marka': 'Nokia', 'model': '3310', 'kolor': 'Czarny'}, {'marka': 'Apple', 'model': '11', 'kolor': 'Złoty'}, {'marka': 'Samsung', 'model': 'Galaxy', 'kolor': 'Srebrny'}]

Sortowanie listy słowników :
[{'marka': 'Nokia', 'model': '3310', 'kolor': 'Czarny'}, {'marka': 'Samsung', 'model': 'Galaxy', 'kolor': 'Srebrny'}, {'marka': 'Apple', 'model': '11', 'kolor': 'Złoty'}]


#### Ćwiczenie

* Napisz program w Pythonie, aby sprawdzić, czy dany ciąg zaczyna się od znaku '`P`', używając `lambda`
* Podpowiedź: skorzystaj z funkcji (metody) `startswith()`

#### Przykładowe rozwiązanie

In [30]:
pass

In [107]:
print(starts_with('Python'))

True


In [108]:
print(starts_with('Java'))

False


#### Ćwiczenie

* Napisz program w Pythonie, aby wyodrębnić rok, miesiąc, dzień i godzinę za pomocą `lambda`
* Podpowiedź: skorzystaj z modułu `datetime`:
  * `now = datetime.datetime.now()` - przypisuje do `now` aktualną lokalną datę i godzinę.
  * `now.year` - wyodrębnia i zwraca rok z `now`.
  * `now.month` - wyodrębnia i zwraca miesiąc z `now`.
  * `now.day` - wyodrębnia i zwraca dzień z `now`.
  * `now.time()` - wyodrębnia i zwraca godzinę z `now`.

#### Przykładowe rozwiązanie

In [109]:
import datetime
now = datetime.datetime.now()
print(now)
pass

2021-09-16 22:16:56.397446
2021
9
16
22:16:56.397446


#### Ćwiczenie

* Napisz program w Pythonie, aby sprawdzić, czy dany ciąg jest liczbą, czy nie, używając `lambda`
* Podpowiedź: przydatna metoda to\
    `string.replace(oldvalue, newvalue, count)`\
    Składnia parametrów:
  * `oldvalue`	– 	Wymagany; ciąg do wyszukania
  * `newvalue`	– 	Wymagany; ciąg znaków, który ma zastąpić starą wartość
  * `count`		– 	Opcjonalny; liczba określająca, ile wystąpień starej wartości chcesz zastąpić; domyślnie są to wszystkie wystąpienia

#### Przykładowe rozwiązanie

In [ ]:
"""Program should work for strings: '26587', '4.2365'"""
pass

## Filtrowanie, aplikowanie, redukowanie

* W Pythonie istnieje kilka przydatnych funkcji korzystających z funkcji anonimowe i pracujących na zbiorach danych:
  * `filter(funkcja, iterator)`: zwraca tylko te wartości ze zbioru danych, dla których `funkcja(wartość)` zwróciła `True`
  * `map(funkcja, iterator)`: aplikuje funkcję dla każdej wartości
  * `reduce(funkcja, iterator)`: służy do zredukowania całego zbioru danych do jednej wartości

### Filtrowanie

* Dla przykładów użyjemy prostego zbioru danych temperatury:

In [13]:
temperatury = [
    37.6, 35.8, 37.6, 33.4, 34.1, 37.1, 35.9, 34.1, 37.1, 40.5, 38.5, 37.6,
    35.8, 34.5, 36.4, 38.3, 37.5, 37.7, 34.0, 35.3, 35.7, 38.9, 34.8, 34.1,
    39.6, 35.4, 34.7, 37.6, 38.4, 36.4, 39.8, 39.1, 37.1, 35.6, 36.8, 37.6,
    36.7, 40.0, 38.0, 34.1, 35.5, 38.5, 36.1, 32.6, 32.9, 34.5, 41.0, 38.3,
    33.7, 38.7, 36.9, 36.2, 33.7, 38.3, 35.3, 38.3, 40.1, 39.3, 38.2, 37.6,
    39.1, 37.1, 34.4, 38.7, 35.8, 38.2, 38.2, 33.1, 37.8, 36.5, 37.6, 37.4,
    34.3, 37.7, 36.0, 37.5, 37.6, 36.5, 31.3, 37.7, 40.3, 39.5, 35.7, 38.1,
    34.7, 36.5, 34.3, 38.0, 37.0, 38.5, 39.4, 37.6, 41.7, 40.0, 38.4, 38.9,
    34.2, 40.2, 34.3, 35.3
]

* Użyjmy teraz funkcji filtrującej, żeby znaleźć wszystkie wartości większe lub równe `40.0`:

In [ ]:
wynik = list(filter(lambda x: x>=40.0,  temperatury))
wynik_sort = sorted(wynik)
print(wynik_sort)

#### Ćwiczenie

* Napisz program w Pythonie do filtrowania listy liczb parzystych i nieparzystych całkowitych za pomocą `lambda` i `filter`

```python
nums = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
```

#### Przykładowe rozwiązanie

In [5]:
nums = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
pass

#### Ćwiczenie

* Napisz program w Pythonie, aby znaleźć przecięcie dwóch podanych tablic za pomocą `lambda` i `filter`

```python
array_nums1 = [1, 2, 3, 5, 7, 8, 9, 10]
array_nums2 = [1, 2, 4, 8, 9]
```

#### Przykładowe rozwiązanie

In [6]:
array_nums1 = [1, 2, 3, 5, 7, 8, 9, 10]
array_nums2 = [1, 2, 4, 8, 9]
pass

#### Ćwiczenie

* Napisz program w Pythonie, aby policzyć parzyste i nieparzyste liczby w danej tablicy liczb całkowitych, używając `lambda` i `filter`

```python
array_nums = [1, 2, 3, 5, 7, 8, 9, 10]
```

#### Przykładowe rozwiązanie

In [7]:
array_nums = [1, 2, 3, 5, 7, 8, 9, 10]
pass

#### Ćwiczenie

* Napisz program w Pythonie, aby znaleźć wartości o długości sześć na podanej liście za pomocą funkcji `lambda` i `filter`

```python
weekdays = ['Poniedziałek', 'Wtorek', 'Środa', 'Czwartek', 'Piątek', 'Sobota', 'Niedziela']
```

#### Przykładowe rozwiązanie

In [9]:
weekdays = ['Poniedziałek', 'Wtorek', 'Środa', 'Czwartek', 'Piątek', 'Sobota', 'Niedziela']
pass

#### Ćwiczenie

* Napisz program w Pythonie, aby znaleźć liczby podzielne przez dziewiętnaście lub trzynaście z listy liczb za pomocą `lambda` i `filter`

```python
nums = [19, 65, 57, 39, 152, 639, 121, 44, 90, 190]
```

#### Przykładowe rozwiązanie

In [ ]:
nums = [19, 65, 57, 39, 152, 639, 121, 44, 90, 190]
pass

#### Ćwiczenie

* Napisz program w Pythonie, aby znaleźć palindromy na podanej liście ciągów za pomocą `lambda` i `filter`
* Palindrom – wyrażenie brzmiące tak samo czytane od lewej do prawej i od prawej do lewej
* Przykładem palindromu jest: „kobyła ma mały bok”

```python
texts = ["php", "w3rr3w", "Python", "abcd", "Java", "aaa"]
```

#### Przykładowe rozwiązanie

In [10]:
texts = ["php", "w3rr3w", "Python", "abcd", "Java", "aaa"]
pass

#### Ćwiczenie

* Napisz program w Pythonie, który zsumuje długość nazw z danej listy nazw po usunięciu nazw zaczynających się od małej litery
* Użyj funkcji `lambda`

```python
sample_names = ['antoni', 'Jakub', 'zuzanna', 'Julia', 'Jan', 'szymon']
```

#### Przykładowe rozwiązanie

In [31]:
sample_names = ['antoni', 'Jakub', 'zuzanna', 'Julia', 'Jan', 'szymon']
pass

### Aplikowanie

In [32]:
print(temperatury)

[37.6, 35.8, 37.6, 33.4, 34.1, 37.1, 35.9, 34.1, 37.1, 40.5, 38.5, 37.6, 35.8, 34.5, 36.4, 38.3, 37.5, 37.7, 34.0, 35.3, 35.7, 38.9, 34.8, 34.1, 39.6, 35.4, 34.7, 37.6, 38.4, 36.4, 39.8, 39.1, 37.1, 35.6, 36.8, 37.6, 36.7, 40.0, 38.0, 34.1, 35.5, 38.5, 36.1, 32.6, 32.9, 34.5, 41.0, 38.3, 33.7, 38.7, 36.9, 36.2, 33.7, 38.3, 35.3, 38.3, 40.1, 39.3, 38.2, 37.6, 39.1, 37.1, 34.4, 38.7, 35.8, 38.2, 38.2, 33.1, 37.8, 36.5, 37.6, 37.4, 34.3, 37.7, 36.0, 37.5, 37.6, 36.5, 31.3, 37.7, 40.3, 39.5, 35.7, 38.1, 34.7, 36.5, 34.3, 38.0, 37.0, 38.5, 39.4, 37.6, 41.7, 40.0, 38.4, 38.9, 34.2, 40.2, 34.3, 35.3]


In [33]:
from statistics import mean
sr_temp = mean(temperatury)
print(sr_temp)

36.9


In [36]:
odch = list(map(lambda x: x - sr_temp, temperatury))
print(odch)

[0.7000000000000028, -1.1000000000000014, 0.7000000000000028, -3.5, -2.799999999999997, 0.20000000000000284, -1.0, -2.799999999999997, 0.20000000000000284, 3.6000000000000014, 1.6000000000000014, 0.7000000000000028, -1.1000000000000014, -2.3999999999999986, -0.5, 1.3999999999999986, 0.6000000000000014, 0.8000000000000043, -2.8999999999999986, -1.6000000000000014, -1.1999999999999957, 2.0, -2.1000000000000014, -2.799999999999997, 2.700000000000003, -1.5, -2.1999999999999957, 0.7000000000000028, 1.5, -0.5, 2.8999999999999986, 2.200000000000003, 0.20000000000000284, -1.2999999999999972, -0.10000000000000142, 0.7000000000000028, -0.19999999999999574, 3.1000000000000014, 1.1000000000000014, -2.799999999999997, -1.3999999999999986, 1.6000000000000014, -0.7999999999999972, -4.299999999999997, -4.0, -2.3999999999999986, 4.100000000000001, 1.3999999999999986, -3.1999999999999957, 1.8000000000000043, 0.0, -0.6999999999999957, -3.1999999999999957, 1.3999999999999986, -1.6000000000000014, 1.399999

#### Format liczb zmiennoprzecinkowych

* **Podczas testów powyższego kodu, Python potrafił zamiast wartości `0.7` zwrócić `0.7000000000000028`**
* **Dlatego dla uproszczonego zapisu wszystkie wartości warto zaokrąglić do jednego miejsca po przecinku**

#### Ćwiczenie

* Napisz program w Pythonie podnoszący do kwadratu i sześcianu(dwie osobne listy) każdą liczbę z podanej listy liczb całkowitych, używając `lambda` i `map`

```python
nums = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
```

#### Przykładowe rozwiązanie

In [ ]:
nums = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
pass

#### Ćwiczenie

* Napisz program w Pythonie, aby dodać dwie podane listy za pomocą `map` i `lambda`

```python
nums1 = [1, 2, 3]
nums2 = [4, 5, 6]
```

#### Przykładowe rozwiązanie

In [19]:
nums1 = [1, 2, 3]
nums2 = [4, 5, 6]
pass

#### Ćwiczenie

* Napisz program w Pythonie, który za pomocą funkcji `lambda` mnoży każdą liczbę z podanej listy przez określoną liczbę
* Wydrukuj wynik

```python
nums = [2, 4, 6, 9, 11]
n = 2
```

#### Przykładowe rozwiązanie

In [18]:
nums = [2, 4, 6, 9, 11]
n = 2
pass

### Redukowanie

* Funkcja `reduce` z modułu `functools` pomaga w iteratywnym redukowaniu zbioru wartości do jednej wartości
* Żeby to lepiej zobrazować, pokażmy dwa przykłady użycia `reduce` do obliczenia:
  * Sumy wartości, i
  * Ich iloczynu

#### Przykład 1: obliczenie sumy

In [20]:
from functools import reduce
nums = [1, 2, 3, 4, 5]
print(reduce(lambda a, b: a + b, nums))

15


* Użycie funkcji `reduce` w powyższym przykładzie sprowadza się tak naprawdę do:

In [ ]:
((((1 + 2) + 3) + 4) + 5)

#### Przykład 2: obliczenie iloczynu

In [ ]:
print(reduce(lambda a, b: a * b, [1, 2, 3, 4]))

* Sprowadza się to do:

In [ ]:
(((1 * 2) * 3) * 4)

#### Ćwiczenie *

* Napisz program w Pythonie do tworzenia ciągu Fibonacciego aż do `n` za pomocą `lambda` i `reduce`
* Podpowiedzi:
  * Da się to zrobić w zasadzie w jednej linijce!
  * Przypomnij sobie, co robiła zmienna `_`

#### Przykładowe rozwiązanie

In [21]:
from functools import reduce
pass
# 1, 1, 2, 3, 5, 8, 13, 21,

In [ ]:
print("Ciąg Fibonacciego do 2:")
print(fib_series(2))

In [ ]:
print("Ciąg Fibonacciego do 5:")
print(fib_series(5))

In [ ]:
print("Ciąg Fibonacciego do 6:")
print(fib_series(6))

In [ ]:
print("Ciąg Fibonacciego do 9:")
print(fib_series(9))

## Tworzenie zbiorów danych w locie

### Wyrażenia listowe – ang. *list comprehension*

* Dzięki generowaniu list w locie możemy:
  * Nawet dość skomplikowane pętle, zamienić na
  * Pojedyncze linijki kodu

* Przykładowo 3 linijki kodu generujące listę sześcianów:

In [ ]:
szesciany = []
for x in range(10):
    szesciany.append(x**3)

In [ ]:
print(szesciany)

* W podstawowej wersji możemy „przenieść” pętlę do nawiasów kwadratowych:

In [ ]:
szesciany = [x**3 for x in range(10)]

In [ ]:
print(szesciany)

* Ogólniejsza postać *list comprehension*:

```python
nowa_lista = [funkcja(element) for element in lista if warunek(element)]
```

* Na przykład by przygotować listę kwadratów liczb nieparzystych z zakresu od `1` do `101`:

In [ ]:
kwadraty = [el**2 for el in range(1, 102) if el % 2 != 0]

In [ ]:
print(kwadraty)

#### Ćwiczenie

* Napisz program w Pythonie, który usuwa liczby dodatnie z podanej listy liczb
* Zsumuj liczby ujemne i wydrukuj wartość bezwzględną za pomocą tworzenia listy – ang. *list comprehension*
* Wydrukuj wynik

```python
nums = [2, 4, -6, -9, 11, -12, 14, -5, 17]
```

#### Przykładowe rozwiązanie

In [ ]:
nums = [2, 4, -6, -9, 11, -12, 14, -5, 17]
pass

#### Ćwiczenie

* Napisz program w Pythonie, aby zmienić kolejność liczb dodatnich i ujemnych w danej tablicy (najpierw wszystkie ujemne, potem wszystkie dodatnie) za pomocą tworzenia listy – ang. *list comprehension*

```python
array_nums = [-1, 2, -3, 5, 7, 8, 9, -10]

```

#### Przykładowe rozwiązanie

In [22]:
array_nums = [-1, 2, -3, 5, 7, 8, 9, -10]
pass

#### Ćwiczenie

* Napisz program w Pythonie, aby:
  * Znaleźć liczby z podanego ciągu
  * Zapisać je na liście
  * Wyświetlić liczby w posortowanej formie
* Użyj funkcji tworzenia listy – ang. *list comprehension*, aby rozwiązać problem

```python
str1 = "sdf 23 safs8 5 sdfsd8 sdfs 56 21sfs 20 5"
```

#### Przykładowe rozwiązanie

In [ ]:
str1 = "sdf 23 safs8 5 sdfsd8 sdfs 56 21sfs 20 5"
pass

### Tworzenie zbiorów

* W podobny sposób można też przygotować zbiór

In [37]:
zbior = {znak for znak in "abracadabra" if znak not in "abc"}

In [38]:
print(zbior)

{'d', 'r'}


### Tworzenie słowników

* Podobnie można stworzyć słownik

In [39]:
tekst = "abracadabra"
wystapienia = {znak: tekst.count(znak) for znak in tekst}

In [40]:
print(wystapienia)

{'a': 5, 'b': 2, 'r': 2, 'c': 1, 'd': 1}


## Wyrażenia generatorowe – ang. *generator expressions*

* Co do zapisu: nie różnią się niczym od *list comprehension*, poza zmianą znaków nawiasów z `[]` na `()`

In [27]:
list_comp = [x ** 0.5 for x in range(1, 11)]
gene_expr = (x ** 0.5 for x in range(1, 11))
# print(list_comp)
# print(gene_expr)

In [24]:
for x in list_comp:
    print(x)

1.0
1.4142135623730951
1.7320508075688772
2.0
2.23606797749979
2.449489742783178
2.6457513110645907
2.8284271247461903
3.0
3.1622776601683795


In [25]:
for x in gene_expr:
    print(x)

1.0
1.4142135623730951
1.7320508075688772
2.0
2.23606797749979
2.449489742783178
2.6457513110645907
2.8284271247461903
3.0
3.1622776601683795


* Różnica polega na tym, że *list comprehension* tworzy listę, a więc zajmuje miejsce w pamięci i czas procesora
* Wyrażenie generatorowe jest dopiero obliczane przy przechodzeniu przez nie, np. w formie pętli `for`

In [41]:
list_comp = [x ** 0.5 for x in range(1, 50000001)]

In [42]:
sum = 0
for x in list_comp:
    sum += x
print(sum)

235702263930.8374


In [43]:
gene_expr = (x ** 0.5 for x in range(1, 50000001))
# gene_expr nie obliczyło jeszcze tych milionów pierwiastków

In [44]:
sum = 0
for x in gene_expr:
    # teraz już kolejne elementy wyrażenia generatorowego są obliczane
    sum += x
print(sum)

235702263930.8374


## Zadania utrwalające

#### Zadanie 1

* Napisz funkcję anonimową obliczającą kwadrat danej wartości
* Czy potrafisz wytłumaczyć czym się to różni od użycia funkcji
```python
pow(a, b)
```
lub operatora
```python
a ** b
```
?

#### Zadanie 2

* Napisz funkcję anonimową obliczającą średnią arytmetyczną dwóch wartości

#### Zadanie 3

* Napisz funkcję anonimową sprawdzającą czy dana wartość jest większa (lub mniejsza) od pewnej wartości
* Taka funkcja anonimowa powinna zwracać wartość logiczną

#### Zadanie 4

* Znajdź wszystkie wartości mniejsze od `36.6`

#### Zadanie 5

* Odfiltruj wszystkie wartości, które są:
  * Mniejsze bądź równe `35.0`, lub 
  * Większe bądź równe `40.0`

#### Zadanie 6

* Oblicz dla każdej liczby w zbiorze temperatur kwadrat:
  * Różnicy tej liczby, i
  * Średniej arytmetycznej całego zbioru

#### Zadanie 7

* Mając obliczone wartości odchyleń temperatur od średniej temperatury, oblicz **wariancję** tych wartości
* Wariancja:
  * Klasyczna **miara zmienności**
  * Intuicyjnie utożsamiana ze zróżnicowaniem zbiorowości
  * Jest **średnią arytmetyczną kwadratów odchyleń** (różnic):
    * Poszczególnych wartości **cechy**, od 
    * **Wartości oczekiwanej**
  * [https://www.matemaks.pl/wariancja.html](https://www.matemaks.pl/wariancja.html)
* Nie zapomnij o podzieleniu przez ilość elementów w zbiorze!

#### Zadanie 8

* Napisz pogram, który policzy kwadraty liczb z zakresu `[1,10000]`, które podzielne są przez:
  * `5`, lub 
  * `9`
* Następnie sprawdź, które z uzyskanych liczb są podzielne:
  * Zarówno przez `5`,
  * Jak i przez `9`

#### Zadanie 9

* Stwórz zbiór, który wypełnisz `30` wywołaniami funkcji:
```python
random.randint(1, 30)
```
* Ile razy wylosowana została ta sama wartość?
* Sprawdzisz to porównując liczbę elementów zbioru.
* Jak się zachowa program, jeśli zmienisz argumenty do funkcji `randint`? 
* A jeśli zmienisz liczbę wywołań tej funkcji?